cunumeric/module.py

```diff
@@ -6227,3 +6236,699 @@ def bincount(
+        if keepdims:
+            remaining_shape = tuple(
+                [
+                    1 if k in axes_set else original_shape[k]
+                    for k in range(0, len(original_shape))
+                ]
+            )
+        else:
+            remaining_shape = tuple(
+                [arr.shape[k] for k in range(0, ndims) if k != axis]
+            )
+
+    # compose qarr.shape with arr.shape:
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # fails with cunumeric:
+    # (numpy works)
+    #
+    # cunumeric BUG: TypeError: a bytes-like object is required, not 'ndarray';
+    # TODO: open nvbug
+    #
+    # qs_all = ndarray(shape = qresult_shape,
+    #                     buffer = array(result_1D),
+    #                     dtype = to_dtype)
+    #
+    # construct result NdArray, non-flattening approach:
+    #
+    if qs_all is None:
+        qs_all = zeros(qresult_shape, dtype=to_dtype)
+    else:
+        # implicit conversion from to_dtype to qs_all.dtype assumed
+        #
+        if qs_all.shape != qresult_shape:
+            raise ValueError("wrong shape on output array")
+
+    for index, q in np.ndenumerate(q_arr):
+        (gamma, j) = method(q, n)
+        (left_pos, right_pos) = (j, j + 1)
+
+        # (N-1) dimensional ndarray of left, right
+        # neighbor values:
+        #
+        # non-flattening approach:
+        #
+        # extract values at index=left_pos;
+        arr_1D_lvals = arr.take(left_pos, axis)
+        arr_vals_shape = arr_1D_lvals.shape
+
+        if right_pos >= n:
+            # some quantile methods may result in j==(n-1),
+            # hence (j+1) could surpass array boundary;
+            #
+            arr_1D_rvals = zeros(arr_vals_shape, dtype=arr_1D_lvals.dtype)
+        else:
+            # extract values at index=right_pos;
+            arr_1D_rvals = arr.take(right_pos, axis)
+
+        # vectorized for axis != None;
+        # (non-flattening approach)
+        #
+        if len(index) == 0:
+            qs_all[...] = (1.0 - gamma) * arr_1D_lvals.reshape(
+                qs_all.shape
+            ) + gamma * arr_1D_rvals.reshape(qs_all.shape)
+        else:
+            qs_all[index] = (1.0 - gamma) * arr_1D_lvals.reshape(
+                qs_all[index].shape
+            ) + gamma * arr_1D_rvals.reshape(qs_all[index].shape)
+
+    return qs_all
+
+
+# Quantiles
+
+
+@add_boilerplate("b")
+def quantile(
+    b: Any,  # FIXME: more narrow mypy-type for scalar, list, tuple, ndarray
+    q: Union[float, Iterable[float], ndarray],
+    axis: Any = None,  # FIXME: add missing mechanism to handle this type decl.
+    out: Optional[ndarray] = None,
+    overwrite_input: Optional[bool] = False,
+    method: str = "linear",
+    keepdims: bool = False,
+) -> ndarray:
+    """
+    Compute the q-th quantile of the data along the specified axis.
+    Parameters
+    ----------
+    b : array_like
```

Comment:
The original code, which typed `a` as simply `ndarray` passes all tests for me, and passes mypy. Are you seeing an issue if you change like so?

```suggestion
@add_boilerplate("a")
def quantile(
    a: ndarray,
    q: Union[float, Iterable[float], ndarray],
    axis: Any = None,  # FIXME: add missing mechanism to handle this type decl.
    out: Optional[ndarray] = None,
    overwrite_input: Optional[bool] = False,
    method: str = "linear",
    keepdims: bool = False,
) -> ndarray:
    """
    Compute the q-th quantile of the data along the specified axis.
    Parameters
    ----------
    a : array_like
```

and also remove `a = asarray(b)` below.

This works because the `@add_boilerplate` decorator handles the special cases of scalars, lists, singleton arrays etc., by converting them to `cunumeric.ndarray` on every function entry. The function body itself can be written assuming the input is already a `cunumeric.ndarray`.

Reply:
Huh! That worked. Okay, done. Thank you.